### ISSUE TITLE
skip_duplicates=True silences all errors

### ISSUE THREAD

**Issue by mspacek** [datajoint/datajoint-python#347]
Now that I have `sql_mode = "TRADITIONAL"` enabled on our server (#344), mysql catches various errors, such as trying to enter values outside of an enum. However, if I insert a row with `skip_duplicates=True`, dj skips trying to insert duplicate rows, but unfortunately, it also skips all other errors that mysql would otherwise raise. This is because `skip_duplicates=True` uses the mysql command:

`INSERT IGNORE ...`

at around line 146 in `base_relation.py`. Apparently, this converts all mysql errors to warnings:

https://stackoverflow.com/questions/548541/insert-ignore-vs-insert-on-duplicate-key-update

so the logic in `BaseRelation.insert()` needs to be changed. Looks like it was last changed in commit dcfa8a2cb9339caa26df3b199ef5514cba7c6319. Previously, at some point in 2015, it looks like there was explicit testing in Python for duplicates, here: https://github.com/datajoint/datajoint-python/pull/149/commits/a13831ffb35c8095e9199494b92c2513623d3336#diff-661ea2dd73d26214fc12cf0606680678R196

I don't know if there's some other way to tell mysql to skip duplicates, while still raising all other errors. As suggested in the above stackoverflow post, maybe `INSERT...ON DUPLICATE KEY UPDATE` is a solution.

**Comment by mspacek** [comment]
BTW, maybe we could do a `try..except IntegrityError` and check if the err number is 1062, which seems to correspond to duplicate insertion attempts. Something like:

```python
from pymysql.err import IntegrityError

try:
    self.insert(rows)
except IntegrityError as err:
    errnum, errmsg = err.args
    if errnum == 1062 and skip_duplicates:
        print(err)
    else:
        raise IntegrityError(errnum, errmsg)
```
However, I guess this error is only raised if strict mode is enabled on the server, further complicating things...

**Comment by mspacek** [comment]
Here's a minimal example with `sql_mode=TRADITIONAL` enabled on the server:

```python

import datajoint as dj

schema = dj.schema('dj_mysql_test', locals())
print("Connected to database: %s" % schema.database)

@schema
class Mouse(dj.Manual):
    definition = '\n'.join([
    "m : varchar(128)", # name
    "----",
    "m_sex : enum('F', 'M', 'unknown')"
    ])

m = Mouse()
m.insert1(dict(m='test1', m_sex='M'))
m.insert1(dict(m='test2', m_sex='blarg'), skip_duplicates=True) # fails silently
m.insert1(dict(m='test2', m_sex='blarg'), skip_duplicates=False) # fails with:
# DataError: (1265, "Data truncated for column 'm_sex' at row 1")

print(m.preview())
print('dj.__version__ == ', dj.__version__)


# output:

'''
*m        m_sex    
+-------+ +-------+
test1     M        
test2              
 (2 tuples)


dj.__version__ ==  0.8.0
'''
```

**Comment by eywalker** [comment]
Hi @mspacek, thanks so much for digging into this deeply! Use of `INSERT ... ON DUPLICATE KEY UPDATE` is unfortunately *not* a choice as that will trigger replacement upon duplicate entry and this violates the principle of data integrity in DataJoint.

Provided that proper functioning of DataJoint depends on `sql_mode=STRICT`, we could catch for `INTEGRITY_ERROR` as you suggested. Furthermore, it should be the case that even when `sql_mode=TRADITIONAL`, attempting to insert an already existing primary key into the table will trigger the error. 

By the way, do you currently still have `sql_mode=TRADITIONAL` enabled on your server? Given that it's catching inappropriate `ENUM` entry, it appears that you have enabled `sql_mode=STRICT`?

**Comment by mspacek** [comment]
Hi @eywalker, as far as I know, there is no `sql_mode=STRICT`, just `STRICT_TRANS_TABLES` and `STRICT_ALL_TABLES`. Setting it to `TRADITIONAL` sets [a lot of modes](https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sqlmode_traditional), including the above two strict modes. This is in mysql 5.5:
```
mysql> select @@sql_mode;
+------------------------------------------------------------------------------------------------------------------------------------------------------+
| @@sql_mode                                                                                                                                           |
+------------------------------------------------------------------------------------------------------------------------------------------------------+
| STRICT_TRANS_TABLES,STRICT_ALL_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,TRADITIONAL,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION |
+------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
I'm currently catching the IntegrityError in our own python code, checking the error number, and if the number is 1062, ignoring it and moving on to the next row to insert. Otherwise, I re-raise the error. Seems to work exactly as I'd like, but it would be more appropriate if were done directly within datajoint.

**Comment by dimitri-yatsenko** [comment]
The desired behavior with `ingore_duplicats=True` is:
1. duplicates are quietly ignored
2. non-duplicate tuples are correctly inserted
3. other errors cause the rollback of the entire insert

The native option, `INSERT IGNORE` meets 1 and 2 but not 3. 

The only solution that meets all three is an explicit check before the insert.  But unless enclosed in a transaction, this transaction does not really guarantee to work correctly may multiply problems in the presence of race conditions.  Therefore, the solution must start a transaction if not in one already, exclude the duplicates from the insert, insert, and commit (if not in transaction initially).

I previously removed the transactionless version of the duplicate check in favor of `INSERT IGNORE` for the sake of simplicity.  If we do implement the duplicate check, it should use a transaction.

**Comment by dimitri-yatsenko** [comment]
Thank @mspacek.

To have the complete correct behavior, I propose the following insert behavior with `skip_duplicates=True`.

If inserting only one tuple, then:
1. Attempt insert
2. Catch and ignore integrity error

If inserting multiple tuples
1. Attempt insert
2. If integrity error:
  a. check if within a transaction.  
  b. if not within a transaction, start a transaction.
  c. check which of the tuples are already in table and exclude them from the insert array
  d. insert
  e. unless initially in transaction, commit

**Comment by dimitri-yatsenko** [comment]
Currently the inserted object can be a generator.  The solution proposed above would require converting into a list to allow the proposed manipulations.

**Comment by dimitri-yatsenko** [comment]
A more general solution with a bit more overhead would be:
1. check if within a transaction.
2. if not within a transaction, start a transaction.
3. while forming the insert query, check which entries are duplicate and exclude them
4. insert
5. unless initially in transaction, commit

The downside of this solution is that it starts a transaction when one is not necessary, e.g. in case of only tuple but it is not too bad.  So probably this is the correct approach.

**Comment by mspacek** [comment]
Sounds good to me, but I'm not familiar enough with transactions, so I can't really comment.

**Comment by dimitri-yatsenko** [comment]
The downside of this solution is that it is inefficient.  The insertion of every tuple requires one extra query, which causes some 20-50 ms network delay when working with a remote server.  This was the reason why we compromised and used `INSERT IGNORE` until now.  In normal practice, users should prevent situations in which duplicates may occur.  Let's implement this correctly but document clearly that `skip_duplicates` can dramatically slow things down and should be used sparingly.  In normal situations like in a properly written and used `_make_tuples`, `skip_duplicates` should not be used.  Duplicates indicate a bug in the application program.

**Comment by dimitri-yatsenko** [comment]
This will be fixed with #382 in a better way than discussed above.
